vault/token_store.go

```diff
@@ -1200,16 +1200,31 @@ func (ts *TokenStore) revokeTree(ctx context.Context, id string) error {
 // Updated to be non-recursive and revoke child tokens
 // before parent tokens(DFS).
 func (ts *TokenStore) revokeTreeSalted(ctx context.Context, saltedID string) error {
-	var dfs []string
-	dfs = append(dfs, saltedID)
+	dfs := []string{saltedID}
+	seenIDs := map[string]struct{}{
+		saltedID: struct{}{},
+	}
 
 	for l := len(dfs); l > 0; l = len(dfs) {
 		id := dfs[0]
 		path := parentPrefix + id + "/"
-		children, err := ts.view.List(ctx, path)
+
+		childrenRaw, err := ts.view.List(ctx, path)
 		if err != nil {
 			return errwrap.Wrapf("failed to scan for children: {{err}}", err)
 		}
+
+		// Filter the child list to remove any items that have ever been in the dfs queue.
+		// This is a robustness check, as a parent/child cycle can lead to an OOM crash.
+		children := make([]string, 0, len(childrenRaw))
+		for _, child := range childrenRaw {
+			if _, seen := seenIDs[child]; !seen {
+				children = append(children, child)
+			} else {
+				ts.Logger().Warn("token cycle found", "token", child)
```

Comment:
It might be best to remove the key that is causing the cycle here, otherwise we will lose all reference to this key and it will live forever

Reply:
If we delete the key here, the depth-first order is bypassed, which could break references if the whole operation doesn't complete. e.g. if ID `root` showed up as a child and we deleted it, revoke fails for some reason, now there are possibly tokens with `root` as an ancestor but `root` is gone.

But as to whether everything gets cleaned up the way it's written, I _think_ it does:

1. If the ID is showing up as as "seen", then it was definitely added to the dfs queue (line 1248).
2. Eventually it was or will be at the front of the queue, with no more of its own children to delete, and it will be deleted.

I'm happy to add tests of any corner cases for this.

Reply:
I think it's worth testing out, I could see a situation where `parentPrefix + id + "/" + child` is never cleaned up if the child token has the parent encoded incorrectly. Which if we are in this situation is likely.

Reply:
Also i wasn't proposing deleting the whole token here, just the extra parent reference that is causing it to be seen more than once. The previously seen version of the token will still allow for it to be cleaned up.

Reply:
Yep, adding a delete of the parent reference here is a good improvement.